File: interview_handwrite/lru_cache_ds.py
"""LRU cache — doubly linked list + hash map."""

from __future__ import annotations

import threading
from typing import Dict, Generic, List, Optional, TypeVar

K = TypeVar("K")
V = TypeVar("V")


class _Node(Generic[K, V]):
    def __init__(self, key: K, value: V):
        self.key = key
        self.value = value
        self.prev: Optional["_Node[K, V]"] = None
        self.next: Optional["_Node[K, V]"] = None


class LRUCache(Generic[K, V]):
    """get/put are O(1)."""

    def __init__(self, capacity: int):
        if capacity <= 0:
            raise ValueError("capacity must be positive")
        self.capacity = capacity
        self._map: Dict[K, _Node[K, V]] = {}

        self._head = _Node(None, None)  # type: ignore[arg-type]
        self._tail = _Node(None, None)  # type: ignore[arg-type]
        self._head.next = self._tail
        self._tail.prev = self._head

    def _remove(self, node: _Node[K, V]) -> None:
        assert node.prev and node.next
        node.prev.next = node.next
        node.next.prev = node.prev

    def _insert_front(self, node: _Node[K, V]) -> None:
        node.next = self._head.next
        node.prev = self._head
        assert self._head.next
        self._head.next.prev = node
        self._head.next = node

    def _touch(self, node: _Node[K, V]) -> None:
        self._remove(node)
        self._insert_front(node)

    def get(self, key: K, default: Optional[V] = None) -> Optional[V]:
        if key not in self._map:
            return default
        node = self._map[key]
        self._touch(node)
        return node.value

    def put(self, key: K, value: V) -> None:
        if key in self._map:
            node = self._map[key]
            node.value = value
            self._touch(node)
            return

        node = _Node(key, value)
        self._map[key] = node
        self._insert_front(node)

        if len(self._map) > self.capacity:
            lru = self._tail.prev
            assert lru is not None and lru is not self._head
            self._remove(lru)
            del self._map[lru.key]

    def __len__(self) -> int:
        return len(self._map)

    def keys_from_mru_to_lru(self) -> List[K]:
        keys: List[K] = []
        cur = self._head.next
        while cur and cur is not self._tail:
            keys.append(cur.key)
            cur = cur.next
        return keys


class ThreadSafeLRUCache(Generic[K, V]):
    def __init__(self, capacity: int):
        self._cache = LRUCache[K, V](capacity)
        self._lock = threading.Lock()

    def get(self, key: K, default: Optional[V] = None) -> Optional[V]:
        with self._lock:
            return self._cache.get(key, default)

    def put(self, key: K, value: V) -> None:
        with self._lock:
            self._cache.put(key, value)

    def __len__(self) -> int:
        with self._lock:
            return len(self._cache)


def _test() -> None:
    cache = LRUCache[int, str](2)
    cache.put(1, "a")
    cache.put(2, "b")
    assert cache.get(1) == "a"
    cache.put(3, "c")
    assert cache.get(2) is None
    assert cache.get(3) == "c"
    assert cache.keys_from_mru_to_lru() == [3, 1]

    safe = ThreadSafeLRUCache[int, int](2)
    safe.put(1, 10)
    assert safe.get(1) == 10
    print("lru_cache_ds: ok")


if __name__ == "__main__":
    _test()

File: interview_handwrite/lru_cache_ds.cpp
// LRU cache — std::list + unordered_map.
//
// For the raw-pointer / custom Node version preferred in systems interviews,
// see lru_cache_raw_list.cpp.
//
// Whiteboard talking points:
// - unordered_map gives O(1) lookup to a list iterator.
// - List order is MRU (front) -> LRU (back).
// - splice() moves a node in O(1) without copying the payload.
// - ThreadSafeLRUCache wraps the core cache with one coarse mutex.

#include <cassert>
#include <iostream>
#include <list>
#include <mutex>
#include <optional>
#include <stdexcept>
#include <unordered_map>
#include <utility>

template <typename K, typename V>
class LRUCache {
 public:
    explicit LRUCache(size_t capacity) : capacity_(capacity) {
        if (capacity == 0) {
            throw std::invalid_argument("capacity must be positive");
        }
    }

    std::optional<V> get(const K& key) {
        auto it = map_.find(key);
        if (it == map_.end()) {
            return std::nullopt;
        }
        touch(it->second);
        return it->second->second;
    }

    void put(const K& key, const V& value) {
        auto it = map_.find(key);
        if (it != map_.end()) {
            it->second->second = value;
            touch(it->second);
            return;
        }

        items_.emplace_front(key, value);
        map_[key] = items_.begin();

        if (map_.size() > capacity_) {
            auto lru = std::prev(items_.end());
            map_.erase(lru->first);
            items_.erase(lru);
        }
    }

    size_t size() const { return map_.size(); }

    std::list<K> keys_from_mru_to_lru() const {
        std::list<K> keys;
        for (const auto& [key, value] : items_) {
            (void)value;
            keys.push_back(key);
        }
        return keys;
    }

 private:
    using Entry = std::pair<K, V>;
    using EntryList = std::list<Entry>;
    using EntryIt = typename EntryList::iterator;

    size_t capacity_;
    EntryList items_;
    std::unordered_map<K, EntryIt> map_;

    void touch(EntryIt it) {
        items_.splice(items_.begin(), items_, it);
    }
};

template <typename K, typename V>
class ThreadSafeLRUCache {
 public:
    explicit ThreadSafeLRUCache(size_t capacity) : cache_(capacity) {}

    std::optional<V> get(const K& key) {
        std::lock_guard<std::mutex> lock(mutex_);
        return cache_.get(key);
    }

    void put(const K& key, const V& value) {
        std::lock_guard<std::mutex> lock(mutex_);
        cache_.put(key, value);
    }

    size_t size() const {
        std::lock_guard<std::mutex> lock(mutex_);
        return cache_.size();
    }

 private:
    LRUCache<K, V> cache_;
    mutable std::mutex mutex_;
};

bool test_lru_cache_ds() {
    LRUCache<int, std::string> cache(2);
    cache.put(1, "a");
    cache.put(2, "b");
    assert(cache.get(1) == "a");
    cache.put(3, "c");
    assert(!cache.get(2).has_value());
    assert(cache.get(3) == "c");

    auto keys = cache.keys_from_mru_to_lru();
    assert(keys.size() == 2);
    assert(keys.front() == 3);
    assert(keys.back() == 1);

    ThreadSafeLRUCache<int, int> safe(2);
    safe.put(1, 10);
    assert(safe.get(1) == 10);
    return true;
}

int main() {
    assert(test_lru_cache_ds());
    std::cout << "lru_cache_ds: ok\n";
    return 0;
}
